### ISSUE TITLE
ensure arch directory exists

### ISSUE THREAD

**Issue by miyagawa** [Perl-Toolchain-Gang/local-lib#3]
(Filed on https://rt.cpan.org/Public/Bug/Display.html?id=86120 ages ago, under Module::Install by mistake I don't know why. Re-filing here)

re: https://github.com/miyagawa/cpanminus/issues/263

When you setup local::lib on the shell, you'll get environment variables such as:

    export PERL5LIB="/Users/miyagawa/perl5/lib/perl5:$PERL5LIB";

suppose you don't have the ~/perl5/lib/perl5 directory yet, and run
perl with that PERL5LIB, your @INC looks like:

  @INC:
    /Users/miyagawa/perl5/lib/perl5
    /Users/miyagawa/.plenv/versions/5.18.0/lib/perl5/site_perl/5.18.0/darwin-2level
    ...

Once you have installed at least one arch-dependent (XS) module, your
@INC becomes:

  @INC:
    /Users/miyagawa/perl5/lib/perl5/darwin-2level
    /Users/miyagawa/perl5/lib/perl5
    /Users/miyagawa/.plenv/versions/5.18.0/lib/perl5/site_perl/5.18.0/darwin-2level
    ...

If you 'use local::lib' in your code, this isn't an issue since local::lib puts archlib in @INC explicitly. With PERL5LIB, perl doesn't put archlib unless the directory exists :/

**Comment by karenetheridge** [comment]
Would it be a problem to add the arch directory to `@INC` before it exists?

**Comment by miyagawa** [comment]
Not a problem at all for perl. But the issue here is that local::lib only sets `PERL5LIB`. Adding an arch directory to `@INC` is completely up to perl itself.

```
> mkdir /tmp/perl5
> PERL5LIB=/tmp/perl5 perl -V        
...
  @INC:
    /tmp/perl5
    /Users/miyagawa/.plenv/versions/5.20.1/lib/perl5/site_perl/5.20.1/darwin-2level
    /Users/miyagawa/.plenv/versions/5.20.1/lib/perl5/site_perl/5.20.1
    /Users/miyagawa/.plenv/versions/5.20.1/lib/perl5/5.20.1/darwin-2level
    /Users/miyagawa/.plenv/versions/5.20.1/lib/perl5/5.20.1
    .

> mkdir /tmp/perl5/darwin-2level
> PERL5LIB=/tmp/perl5 perl -V        
...
  @INC:
    /tmp/perl5/darwin-2level
    /tmp/perl5
    /Users/miyagawa/.plenv/versions/5.20.1/lib/perl5/site_perl/5.20.1/darwin-2level
    /Users/miyagawa/.plenv/versions/5.20.1/lib/perl5/site_perl/5.20.1
    /Users/miyagawa/.plenv/versions/5.20.1/lib/perl5/5.20.1/darwin-2level
    /Users/miyagawa/.plenv/versions/5.20.1/lib/perl5/5.20.1
    .
```

**Comment by haarg** [comment]
Note that this is only an issue when modules are loaded mid way through a perl process running.  I think just creating the directory is reasonable though.

**Comment by miyagawa** [comment]
Correct - cpanm is a very special use case for this because it installs modules in the middle of the same perl process :) And so cpanm has a workaround for it.

**Comment by shadowcat-mst** [comment]
this is why local::lib used to put the arch dir in PERL5LIB explicitly; I think somebody insisted on removing that to "clean things up".

We should either revert the behaviour or create the directory.

**Comment by miyagawa** [comment]
> this is why local::lib used to put the arch dir in PERL5LIB explicitly; I think somebody insisted on removing that to "clean things up".

Sounds right. https://github.com/Perl-Toolchain-Gang/local-lib/commit/cdd450c566dcfefb6ee4f0cc3c6255e0c289f854 fixes [RT 67871](https://rt.cpan.org/Public/Bug/Display.html?id=67871)

> We should either revert the behaviour or create the directory.

+1 either would be fine for me.

**Comment by karenetheridge** [comment]
+1 to either one from me too.

**Comment by dolmen** [comment]
As the requestor of RT 67871, I'm in favor of creating the directories.

If creating the directories fails because of filesystem permissions (for example if a readonly local::lib dir is used), just ignore the failure, because CPAN clients will not be more able to write there.

**Comment by shadowcat-mst** [comment]
we used to scaffold a bunch of stuff anyway, mostly so we had somewhere to write .modulebuildrc - but when PERL_MB_OPT came around, we stopped doing that.

the thing is, people can setup a local::lib with intent to -use- it but no expectation that the filesystem will be modified until they ask for it

cpanminus knows it's planning to write though - so maybe we need a --scaffold-directories option to the local::lib import that it can pass to explicitly ask for that?

**Comment by karenetheridge** [comment]
> the thing is, people can setup a local::lib with intent to -use- it but no expectation that the filesystem will be modified until they ask for it

I don't understand -- if I do `perl -Mlocal::lib=foo -e1', the `foo` directory is created. But you're saying the intent is that it should not be?

**Comment by shadowcat-mst** [comment]
if we're creating the 'foo' directory, we might as well create foo/lib/perl5 and foo/lib/perl5/$arch

I'm sort of wondering if we should have a 'create everything' -and- a 'create nothing' mode, but maybe the latter is just 'use lib'

**Comment by shadowcat-mst** [comment]
@miyagawa you mean reverting that commit fixes it?

@dolmen heh, I note you're the one who removed the arch dir in the first place

@haarg thoughts on maybe creating lib/perl5, arch and version dirs (if we're doing version dirs), and maybe bin as well? given we're already creating the root, producing a skeleton tree in there might actually be more DWIM

**Comment by haarg** [comment]
I think we should create the appropriate directories under the local lib dir.  I have some code in progress for this.  We already have a --no-create option to prevent creating the local lib directory.

**Comment by ribasushi** [comment]
> I think we should create the appropriate directories under the local lib dir.

+1

**Comment by miyagawa** [comment]
> if we're creating the 'foo' directory, we might as well create foo/lib/perl5 and foo/lib/perl5/$arch

That's what I meant with the original ticket here, yes.

**Comment by miyagawa** [comment]
Fixed with 2f7bd47d1098d032de6d5f61354e310ad41b1097 I think.
